neps/nep-0508.md

```diff
@@ -0,0 +1,264 @@
+* For the duration of the epoch before the new shard layout takes effect, the node will need to maintain the state and flat state of shards in the old and new layout at the same time. The State and FlatState columns will grow up to approx 2x the size. The processing overhead should be minimal as the chunks will still be executed only on the parent shards. There will be increased load on the database while applying changes to both the parent and the children shards. 
+
+### Resharding flow
+
+* The new shard layout will be agreed on offline by the protocol team and hardcoded in the reference implementation.
+* In epoch T, past the protocol version upgrade date, nodes will vote to switch to the new protocol version. The new protocol version will contain the new shard layout. 
+* In epoch T, in the last block of the epoch, the EpochConfig for epoch T+2 will be set. The EpochConfig for epoch T+2 will have the new shard layout.
+* In epoch T + 1, all nodes will perform the state split. The child shards will be kept up to date with the blockchain up until the epoch end first via catchup, and later as part of block postprocessing state application.
+* In epoch T + 2, the chain will switch to the new shard layout. 
+
+## Reference Implementation
+
+The implementation heavily re-uses the implementation from [NEP-40](https://github.com/near/NEPs/blob/master/specs/Proposals/0040-split-states.md). Below are listed the major differences and additions. 
+
+### Code pointers to the proposed implementation 
+
+* [new shard layout](https://github.com/near/nearcore/blob/c9836ab5b05c229da933d451fe8198d781f40509/core/primitives/src/shard_layout.rs#L161)  
+* [the main logic for splitting states](https://github.com/near/nearcore/blob/c9836ab5b05c229da933d451fe8198d781f40509/chain/chain/src/resharding.rs#L280)
+* [the main logic for applying chunks to split states](https://github.com/near/nearcore/blob/c9836ab5b05c229da933d451fe8198d781f40509/chain/chain/src/update_shard.rs#L315)
+* [the main logic for garbage collecting state from parent shard](https://github.com/near/nearcore/blob/c9836ab5b05c229da933d451fe8198d781f40509/chain/chain/src/store.rs#L2335)
+
+### Flat Storage
+
+The old implementation of resharding relied on iterating over the full trie state of the parent shard in order to build the state for the children shards. This implementation was suitable at the time but since then the state has grown considerably and this implementation is now too slow to fit within a single epoch. The new implementation relies on iterating through the flat storage in order to build the children shards quicker. Based on benchmarks, splitting the largest shard by using flat storage can take around 15 min without throttling and around 3 hours with throttling to maintain the block production rate.
```

Comment:
Unrelated to this proposal, but do I understand correctly that once we have in-memory trie, the old approach would actually work again?

Reply:
Hmm good question. One requirement that we have for storage is to give us access to a frozen view of the chain at a pre-defined block. This is why we need the flat store snapshots. I'm not sure if such snapshottting would be feasible with mem-trie. It's definitely worth exploring as a number of fundamental assumptions will change.

Reply:
Although theoretically it does seem plausible given we have a large enough GC period for mem-trie and we maintain "historic" state for the GC period.

Reply:
> One requirement that we have for storage is to give us access to a frozen view of the chain at a pre-defined block. This is why we need the flat store snapshots. I'm not sure if such snapshottting would be feasible with mem-trie

My understanding is that in-memory trie just moves trie to be held in memory and does not change its persistence assumptions, so yes you can read the state at some predefined block (assuming it has not been garbage collected).

Reply:
In-memory trie loads from flat storage and indeed does not change any persistence assumptions. If there is flat storage available at some block (such as via a flat state snapshot) then memtries can be loaded that corresponds to that flat state. It's possible to then iterate through the memtrie to reconstruct a resharded on-disk trie or flat state. That being said, this does not sound so much like "the old approach works again", but rather that memtries can be used as an optimization to the (pretty isolated) step of reconstructing a trie from flat state -- which is the solution we're adopting in this proposal. We would be doing the same thing, except that memtrie construction from flat state is a bit more efficient than direct trie reconstruction of flat state (for reasons of doing less wasted work and doing less disk I/O).

The pros would be that this is more efficient (perhaps 2x-4x more would be my ballpark guess), the cons is that this requires more memory and, well, we don't have the code for dumping memtrie into on-disk trie. It sounds easy, but they are different data structures and in practice we have to carefully deal with refcount keeping which will not be trivial.